File: java/src/main/java/vsdk/toolkit/raycasting/bidirectionalRaytracing/LightList.java
/**
Definition of the light list class
this class can be used for sampling lights
*/

package vsdk.toolkit.raycasting.bidirectionalRaytracing;

import java.util.ArrayList;

import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.common.logging.Logger;
import vsdk.toolkit.common.dataStructures.CircularList;
import vsdk.toolkit.common.dataStructures.CircularListIterator;
import vsdk.toolkit.common.linealAlgebra.Numeric;
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.material.XxdfComponentFlag;
import vsdk.toolkit.numericalAnalysis.PatchVisitor;
import vsdk.toolkit.environment.geometry.elements.Patch;
import vsdk.toolkit.environment.geometry.elements.RayHit;

public class LightList extends CircularList<LightInfo> {
    // Total flux ( sum(L * A * PI))
    private float totalFlux;
    private float totalImp;
    private boolean includeVirtual;
    private int lightCount;

    // Iteration over lights, not multi-thread!

    // Discrete sampling of light sources

    // A getPatchList must be supplied for building a light list.
    // Non emitting patches (edf == nullptr) are NOT put in the list.
    public LightList(ArrayList<Patch> list, boolean includeVirtualPatches) {
        LightInfo info = new LightInfo();
        ColorRgb lightColor;

        totalFlux = 0.0f;
        lightCount = 0;
        includeVirtual = includeVirtualPatches;
        totalImp = 0.0f;

        for ( int i = 0; list != null && i < list.size(); i++ ) {
            Patch light = list.get(i);
            if ( (!light.hasZeroVertices() || includeVirtual)
                && ( light.material.getEdf() != null ) ) {
                info = new LightInfo();
                info.light = light;

                // calc emittedFlux
                if ( light.hasZeroVertices() ) {
                    ColorRgb e;
                    if ( light.material.getEdf() == null ) {
                        e = new ColorRgb();
                        e.clear();
                    } else {
                        e = light.material.getEdf().phongEmittance((RayHit)null, XxdfComponentFlag.DIFFUSE_COMPONENT);
                    }
                    info.emittedFlux = e.average();
                } else {
                    lightColor = PatchVisitor.averageEmittance(light, XxdfComponentFlag.DIFFUSE_COMPONENT);
                    info.emittedFlux = lightColor.average() * light.area;
                }

                totalFlux += info.emittedFlux;
                lightCount++;
                append(info);
            }
        }
    }

    public LightList(ArrayList<Patch> list) {
        this(list, false);
    }

    public CircularList<LightInfo> entries() {
        return this;
    }

    /**
    Returns sampled patch, scales x_1 back to a random in 0..1
    */
    public Patch sample(double[] x1, double[] pdf) {
        LightInfo info;
        LightInfo lastInfo = null;
        CircularListIterator<LightInfo> iterator = new CircularListIterator<>(this);

        double rnd = x1[0] * totalFlux;
        double currentSum;

        info = iterator.nextOnSequence();
        while ( (info != null) && (info.light.hasZeroVertices()) && (!includeVirtual) ) {
            info = iterator.nextOnSequence();
        }

        if ( info == null ) {
            Logger.warning("CLightList::sample", "No lights available");
            return null;
        }

        currentSum = info.emittedFlux;

        while ( (rnd > currentSum) && (info != null) ) {
            lastInfo = info;
            info = iterator.nextOnSequence();
            while ( (info != null) && (info.light.hasZeroVertices()) && (!includeVirtual) ) {
                info = iterator.nextOnSequence();
            }

            if ( info != null ) {
                currentSum += info.emittedFlux;
            } else {
                info = lastInfo;
                rnd = currentSum - 1.0; // :-(  Damn float inaccuracies
            }
        }

        if ( info != null ) {
            x1[0] = ((x1[0] - ((currentSum - info.emittedFlux) / totalFlux)) /
                (info.emittedFlux / totalFlux));
            if ( pdf != null && pdf.length > 0 ) {
                pdf[0] = info.emittedFlux / totalFlux;
            }
            return info.light;
        }

        return null;
    }

    private double evalPdfVirtual(Patch light, Vector3D point) {
        // EvalPDF for virtual patches (see EvalPDF)
        double probabilityDensityFunction;

        // Prob for choosing this light
        int all = XxdfComponentFlag.DIFFUSE_COMPONENT | XxdfComponentFlag.GLOSSY_COMPONENT | XxdfComponentFlag.SPECULAR_COMPONENT;

        ColorRgb e;
        if ( light.material.getEdf() == null ) {
            e = new ColorRgb();
            e.clear();
        } else {
            e = light.material.getEdf().phongEmittance((RayHit)null, all);
        }
        probabilityDensityFunction = e.average() / totalFlux;

        return probabilityDensityFunction;
    }

    private double evalPdfReal(Patch light, Vector3D point) {
        // Eval PDF for normal patches (see EvalPDF)
        ColorRgb color;
        double pdf;

        color = PatchVisitor.averageEmittance(light, XxdfComponentFlag.DIFFUSE_COMPONENT);

        // Prob for choosing this light
        pdf = color.average() * light.area / totalFlux;

        return pdf;
    }

    public double evalPdf(Patch light, Vector3D point) {
        // TODO!!!  1) patch should become class
        //          2) virtual patch should become child-class
        //          3) this method should be handled by specialisation
        if ( totalFlux < Numeric.EPSILON ) {
            return 0.0;
        }
        if ( light.hasZeroVertices() ) {
            return evalPdfVirtual(light, point);
        } else {
            return evalPdfReal(light, point);
        }
    }

    /* Important light sampling */

    public static double
    computeOneLightImportanceVirtual(
        Patch light,
        Vector3D point,
        Vector3D normal,
        float emittedFlux)
    {
        // ComputeOneLightImportance for virtual patches
        int all = XxdfComponentFlag.DIFFUSE_COMPONENT | XxdfComponentFlag.GLOSSY_COMPONENT | XxdfComponentFlag.SPECULAR_COMPONENT;

        ColorRgb e;

        if ( light.material.getEdf() == null ) {
            e = new ColorRgb();
            e.clear();
        } else {
            e = light.material.getEdf().phongEmittance((RayHit)null, all);
        }
        return e.average();
    }

    public static double
    computeOneLightImportanceReal(
        Patch light,
        Vector3D point,
        Vector3D normal,
        float emittedFlux)
    {
        // ComputeOneLightImportance for real patches
        int tried = 0;  // No positions on the patch are tried yet
        boolean done = false;
        double contribution = 0.0;
        Vector3D lightPoint = new Vector3D();
        Vector3D lightNormal = new Vector3D();
        Vector3D dir = new Vector3D();
        double cosRayPatch;
        double cosRayLight;
        double dist2;

        while ( !done && tried <= light.numberOfVertices ) {
            // Choose a point on the patch according to 'tried'

            if ( tried == 0 ) {
                lightPoint.copy(light.midPoint);
                lightNormal.copy(light.normal);
            } else {
                lightPoint.copy(light.vertex[tried - 1].point);
                if ( light.vertex[tried - 1].normal != null ) {
                    lightNormal.copy(light.vertex[tried - 1].normal);
                } else {
                    lightNormal.copy(light.normal);
                }
            }

            // Estimate the contribution

            // Ray direction (but no ray is shot of course)
            Vector3D copy = new Vector3D(point.x, point.y, point.z);

            dir.subtraction(lightPoint, copy);
            dist2 = dir.norm2();

            // Cosines have an addition distance length in them
            cosRayLight = -dir.dotProduct(lightNormal);
            cosRayPatch = dir.dotProduct(normal);

            if ( cosRayLight > 0 && cosRayPatch > 0 ) {
                // Orientation of surfaces ok.
                // BRDF is not taken into account since
                // we're expecting diffuse/glossy surfaces here.

                contribution = (cosRayPatch * cosRayLight * emittedFlux / (Math.PI * dist2));
                done = true;
            }

            tried++; // Try next point on light
        }

        return contribution;
    }

    public static double
    computeOneLightImportance(
        Patch light,
        Vector3D point,
        Vector3D normal,
        float emittedFlux)
    {
        // TODO!!!  1) patch should become class
        //          2) virtual patch should become child-class
        //          3) this method should be handled by specialisation
        if ( light.hasZeroVertices() ) {
            return computeOneLightImportanceVirtual(light, point, normal, emittedFlux);
        } else {
            return computeOneLightImportanceReal(light, point, normal, emittedFlux);
        }
    }

    private Vector3D lastPoint = new Vector3D();
    private Vector3D lastNormal = new Vector3D();

    public void computeLightImportance(Vector3D point, Vector3D normal) {
        if ( (point.equals(lastPoint, Numeric.EPSILON_FLOAT)) &&
            (normal.equals(lastNormal, Numeric.EPSILON_FLOAT)) ) {
            return; // Still ok !!
        }


        LightInfo info;
        CircularListIterator<LightInfo> iterator = new CircularListIterator<>(this);
        double imp;

        totalImp = 0.0f;

        // next
        info = iterator.nextOnSequence();
        while ( (info != null) && (info.light.hasZeroVertices()) && (!includeVirtual) ) {
            info = iterator.nextOnSequence();
        }

        while ( info != null ) {
            imp = computeOneLightImportance(info.light, point, normal, info.emittedFlux);
            totalImp += (float)imp;
            info.importance = (float)imp;

            // next
            info = iterator.nextOnSequence();
            while ( (info != null) && (info.light.hasZeroVertices()) && (!includeVirtual) ) {
                info = iterator.nextOnSequence();
            }
        }
    }

    public Patch sampleImportant(Vector3D point, Vector3D normal, double[] x1, double[] pdf) {
        LightInfo info;
        LightInfo lastInfo = null;
        CircularListIterator<LightInfo> iterator = new CircularListIterator<>(this);
        double rnd;
        double currentSum;

        computeLightImportance(point, normal);

        if ( totalImp == 0 ) {
            // No light is important, but we must return one (->optimize ?)
            return sample(x1, pdf);
        }

        rnd = x1[0] * totalImp;

        // Next
        info = iterator.nextOnSequence();
        while ( (info != null) && (info.light.hasZeroVertices()) && (!includeVirtual) ) {
            info = iterator.nextOnSequence();
        }

        if ( info == null ) {
            Logger.warning("CLightList::sample", "No lights available");
            return null;
        }

        currentSum = info.importance;

        while ( (rnd > currentSum) && (info != null) ) {
            lastInfo = info;

            // next
            info = iterator.nextOnSequence();
            while ( info != null && info.light.hasZeroVertices() && !includeVirtual ) {
                info = iterator.nextOnSequence();
            }

            if ( info != null ) {
                currentSum += info.importance;
            } else {
                info = lastInfo;
                rnd = currentSum - 1.0; // :-(  Damn float inaccuracies
            }
        }

        if ( info != null ) {
            x1[0] = ((x1[0] - ((currentSum - info.importance) / totalImp)) /
                (info.importance / totalImp));
            if ( pdf != null && pdf.length > 0 ) {
                pdf[0] = info.importance / totalImp;
            }
            return info.light;
        }

        return null;
    }

    public double
    evalPdfImportant(
        Patch light,
        Vector3D lightPoint,
        Vector3D litPoint,
        Vector3D normal)
    {
        double pdf;
        LightInfo info;
        CircularListIterator<LightInfo> iterator = new CircularListIterator<>(this);

        computeLightImportance(litPoint, normal);

        // Search the light in the list :-(

        do {
            info = iterator.nextOnSequence();
        } while ( info != null && info.light != light );

        if ( info == null ) {
            Logger.warning("CLightList::evalPdfImportant", "Could not find light");
            return 0.0;
        }

        // Prob for choosing this light
        if ( totalImp < Numeric.EPSILON ) {
            pdf = 0.0;
        } else {
            pdf = info.importance / totalImp;
        }

        return pdf;
    }
}

File: typescript/src/main/vsdk/toolkit/raycasting/bidirectionalRaytracing/LightList.ts
import { ArrayList } from "../../../../java/util/ArrayList";
import { ColorRgb } from "../../common/color/ColorRgb";
import { Logger as VsdkLogger } from "../../common/logging/Logger";
import { CircularList } from "../../common/dataStructures/CircularList";
import { CircularListIterator } from "../../common/dataStructures/CircularListIterator";
import { Numeric } from "../../common/linealAlgebra/Numeric";
import { Vector3D } from "../../common/linealAlgebra/Vector3D";
import { XxdfComponentFlag } from "../../material/XxdfComponentFlag";
import { PatchVisitor } from "../../numericalAnalysis/PatchVisitor";
import { Patch } from "../../environment/geometry/elements/Patch";
import { LightInfo } from "./LightInfo";

export class LightList extends CircularList<LightInfo> {
  private totalFlux: number;
  private totalImp: number;
  private includeVirtual: boolean;
  private lightCount: number;

  private lastPoint: Vector3D;
  private lastNormal: Vector3D;

  public constructor(list: ArrayList<Patch>, includeVirtualPatches = false) {
    super();

    this.totalFlux = 0.0;
    this.lightCount = 0;
    this.includeVirtual = includeVirtualPatches;
    this.totalImp = 0.0;

    this.lastPoint = new Vector3D();
    this.lastNormal = new Vector3D();

    for (let i = 0; list !== null && i < list.size(); i++) {
      const light = list.get(i);
      const hasEdf = light.material !== null && light.material.getEdf() !== null;
      if ((!light.hasZeroVertices() || this.includeVirtual) && hasEdf) {
        const info = new LightInfo();
        info.light = light;

        if (light.hasZeroVertices()) {
          let e = new ColorRgb();
          if (light.material === null || light.material.getEdf() === null) {
            e.clear();
          }
          else {
            e = light.material.getEdf()!.phongEmittance(null as any, XxdfComponentFlag.DIFFUSE_COMPONENT);
          }
          info.emittedFlux = e.average();
        }
        else {
          const lightColor = PatchVisitor.averageEmittance(light, XxdfComponentFlag.DIFFUSE_COMPONENT);
          info.emittedFlux = lightColor.average() * light.area;
        }

        this.totalFlux += info.emittedFlux;
        this.lightCount++;
        this.append(info);
      }
    }
  }

  public entries(): CircularList<LightInfo> {
    return this;
  }

  public sample(x1: number[], pdf: number[] | null): Patch | null {
    let lastInfo: LightInfo | null = null;
    const iterator = new CircularListIterator<LightInfo>(this);

    const rnd = x1[0] * this.totalFlux;

    let info = iterator.nextOnSequence();
    while (info !== null && info.light !== null && info.light.hasZeroVertices() && !this.includeVirtual) {
      info = iterator.nextOnSequence();
    }

    if (info === null) {
      VsdkLogger.warning("CLightList::sample", "No lights available");
      return null;
    }

    let currentSum = info.emittedFlux;

    while (rnd > currentSum && info !== null) {
      lastInfo = info;
      info = iterator.nextOnSequence();
      while (info !== null && info.light !== null && info.light.hasZeroVertices() && !this.includeVirtual) {
        info = iterator.nextOnSequence();
      }

      if (info !== null) {
        currentSum += info.emittedFlux;
      }
      else {
        info = lastInfo;
        currentSum = currentSum - 1.0;
      }
    }

    if (info !== null) {
      x1[0] = (x1[0] - ((currentSum - info.emittedFlux) / this.totalFlux))
        / (info.emittedFlux / this.totalFlux);
      if (pdf !== null && pdf.length > 0) {
        pdf[0] = info.emittedFlux / this.totalFlux;
      }
      return info.light;
    }

    return null;
  }

  private evalPdfVirtual(light: Patch, point: Vector3D): number {
    void point;
    let probabilityDensityFunction: number;

    const all = XxdfComponentFlag.DIFFUSE_COMPONENT | XxdfComponentFlag.GLOSSY_COMPONENT | XxdfComponentFlag.SPECULAR_COMPONENT;

    let e: ColorRgb;

    if (light.material === null || light.material.getEdf() === null) {
      e = new ColorRgb();
      e.clear();
    }
    else {
      e = light.material.getEdf()!.phongEmittance(null as any, all);
    }
    probabilityDensityFunction = e.average() / this.totalFlux;

    return probabilityDensityFunction;
  }

  private evalPdfReal(light: Patch, point: Vector3D): number {
    void point;
    const color = PatchVisitor.averageEmittance(light, XxdfComponentFlag.DIFFUSE_COMPONENT);

    return color.average() * light.area / this.totalFlux;
  }

  public evalPdf(light: Patch, point: Vector3D): number {
    if (this.totalFlux < Numeric.EPSILON) {
      return 0.0;
    }
    if (light.hasZeroVertices()) {
      return this.evalPdfVirtual(light, point);
    }
    return this.evalPdfReal(light, point);
  }

  public static computeOneLightImportanceVirtual(
    light: Patch,
    point: Vector3D,
    normal: Vector3D,
    emittedFlux: number
  ): number {
    void point;
    void normal;
    void emittedFlux;
    const all = XxdfComponentFlag.DIFFUSE_COMPONENT | XxdfComponentFlag.GLOSSY_COMPONENT | XxdfComponentFlag.SPECULAR_COMPONENT;

    let e: ColorRgb;

    if (light.material === null || light.material.getEdf() === null) {
      e = new ColorRgb();
      e.clear();
    }
    else {
      e = light.material.getEdf()!.phongEmittance(null as any, all);
    }
    return e.average();
  }

  public static computeOneLightImportanceReal(
    light: Patch,
    point: Vector3D,
    normal: Vector3D,
    emittedFlux: number
  ): number {
    let tried = 0;
    let done = false;
    let contribution = 0.0;
    const lightPoint = new Vector3D();
    const lightNormal = new Vector3D();
    const dir = new Vector3D();

    while (!done && tried <= light.numberOfVertices) {
      if (tried === 0) {
        lightPoint.copy(light.midPoint);
        lightNormal.copy(light.normal);
      }
      else {
        lightPoint.copy(light.vertex[tried - 1]!.point);
        if (light.vertex[tried - 1]!.normal !== null) {
          lightNormal.copy(light.vertex[tried - 1]!.normal as Vector3D);
        }
        else {
          lightNormal.copy(light.normal);
        }
      }

      const copy = new Vector3D(point.x, point.y, point.z);

      dir.subtraction(lightPoint, copy);
      const dist2 = dir.norm2();

      const cosRayLight = -dir.dotProduct(lightNormal);
      const cosRayPatch = dir.dotProduct(normal);

      if (cosRayLight > 0 && cosRayPatch > 0) {
        contribution = cosRayPatch * cosRayLight * emittedFlux / (globalThis.Math.PI * dist2);
        done = true;
      }

      tried++;
    }

    return contribution;
  }

  public static computeOneLightImportance(
    light: Patch,
    point: Vector3D,
    normal: Vector3D,
    emittedFlux: number
  ): number {
    if (light.hasZeroVertices()) {
      return LightList.computeOneLightImportanceVirtual(light, point, normal, emittedFlux);
    }
    return LightList.computeOneLightImportanceReal(light, point, normal, emittedFlux);
  }

  public computeLightImportance(point: Vector3D, normal: Vector3D): void {
    if (point.equals(this.lastPoint, Numeric.EPSILON_FLOAT)
      && normal.equals(this.lastNormal, Numeric.EPSILON_FLOAT)) {
      return;
    }

    const iterator = new CircularListIterator<LightInfo>(this);

    this.totalImp = 0.0;

    let info = iterator.nextOnSequence();
    while (info !== null && info.light !== null && info.light.hasZeroVertices() && !this.includeVirtual) {
      info = iterator.nextOnSequence();
    }

    while (info !== null) {
      const imp = LightList.computeOneLightImportance(info.light as Patch, point, normal, info.emittedFlux);
      this.totalImp += imp;
      info.importance = imp;

      info = iterator.nextOnSequence();
      while (info !== null && info.light !== null && info.light.hasZeroVertices() && !this.includeVirtual) {
        info = iterator.nextOnSequence();
      }
    }

    this.lastPoint.copy(point);
    this.lastNormal.copy(normal);
  }

  public sampleImportant(point: Vector3D, normal: Vector3D, x1: number[], pdf: number[] | null): Patch | null {
    let lastInfo: LightInfo | null = null;
    const iterator = new CircularListIterator<LightInfo>(this);

    this.computeLightImportance(point, normal);

    if (this.totalImp === 0) {
      return this.sample(x1, pdf);
    }

    const rnd = x1[0] * this.totalImp;

    let info = iterator.nextOnSequence();
    while (info !== null && info.light !== null && info.light.hasZeroVertices() && !this.includeVirtual) {
      info = iterator.nextOnSequence();
    }

    if (info === null) {
      VsdkLogger.warning("CLightList::sample", "No lights available");
      return null;
    }

    let currentSum = info.importance;

    while (rnd > currentSum && info !== null) {
      lastInfo = info;

      info = iterator.nextOnSequence();
      while (info !== null && info.light !== null && info.light.hasZeroVertices() && !this.includeVirtual) {
        info = iterator.nextOnSequence();
      }

      if (info !== null) {
        currentSum += info.importance;
      }
      else {
        info = lastInfo;
        currentSum = currentSum - 1.0;
      }
    }

    if (info !== null) {
      x1[0] = (x1[0] - ((currentSum - info.importance) / this.totalImp))
        / (info.importance / this.totalImp);
      if (pdf !== null && pdf.length > 0) {
        pdf[0] = info.importance / this.totalImp;
      }
      return info.light;
    }

    return null;
  }

  public evalPdfImportant(
    light: Patch,
    lightPoint: Vector3D,
    litPoint: Vector3D,
    normal: Vector3D
  ): number {
    void lightPoint;
    this.computeLightImportance(litPoint, normal);

    const iterator = new CircularListIterator<LightInfo>(this);

    let info: LightInfo | null;
    do {
      info = iterator.nextOnSequence();
    } while (info !== null && info.light !== light);

    if (info === null) {
      VsdkLogger.warning("CLightList::evalPdfImportant", "Could not find light");
      return 0.0;
    }

    let pdf: number;
    if (this.totalImp < Numeric.EPSILON) {
      pdf = 0.0;
    }
    else {
      pdf = info.importance / this.totalImp;
    }

    return pdf;
  }
}
